File: src/uk/co/ben_gibson/open/in/git/host/UI/Action/Action.java
package uk.co.ben_gibson.open.in.git.host.UI.Action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import uk.co.ben_gibson.open.in.git.host.Container;
import uk.co.ben_gibson.open.in.git.host.Exception.OpenInGitHostException;
import uk.co.ben_gibson.open.in.git.host.Logger.Logger;

abstract class Action extends AnAction
{
    protected abstract boolean shouldActionBeEnabled(AnActionEvent event);

    abstract void actionPerformed(Project project, AnActionEvent event) throws OpenInGitHostException;

    public void actionPerformed(AnActionEvent event)
    {
        Project project = event.getProject();

        Logger logger = container().logger(project);

        if (project == null) {
            return;
        }

        try {
            this.actionPerformed(project, event);
        } catch (OpenInGitHostException exception) {
            logger.exception(exception);
        }
    }

    public void update(AnActionEvent event)
    {
        super.update(event);

        if (event.getProject() == null) {
            event.getPresentation().setEnabled(false);
            return;
        }

        event.getPresentation().setEnabledAndVisible(this.shouldActionBeEnabled(event));

        if (event.getPresentation().isEnabled()) {
            event.getPresentation().setIcon(this.container().remoteHost(event.getProject()).icon());
        }
    }

    Container container()
    {
        return ServiceManager.getService(Container.class);
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/ui/actions/Action.kt
package uk.co.ben_gibson.git.link.ui.actions

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.components.service
import com.intellij.openapi.project.DumbAwareAction
import com.intellij.openapi.project.Project
import uk.co.ben_gibson.git.link.Context
import uk.co.ben_gibson.git.link.GitLinkBundle
import uk.co.ben_gibson.git.link.copyToClipBoard
import uk.co.ben_gibson.git.link.openInBrowser
import uk.co.ben_gibson.git.link.platform.PlatformLocator

abstract class Action(private val type: Type): DumbAwareAction() {

    enum class Type(val key: String) {
        BROWSER("browser"),
        COPY("copy"),
        COPY_MARKDOWN("copy-markdown")
    }

    abstract fun buildContext(project: Project, event: AnActionEvent) : Context?

    override fun actionPerformed(event: AnActionEvent) {
        val project = event.project ?: return

        val context = buildContext(project, event) ?: return

        when(type) {
            Type.BROWSER -> openInBrowser(project, context)
            Type.COPY -> copyToClipBoard(project, context)
            Type.COPY_MARKDOWN -> copyToClipBoard(project, context, true)
        }
    }

    open fun shouldBeEnabled(event: AnActionEvent) = true

    override fun update(event: AnActionEvent) {
        super.update(event)

        event.presentation.isEnabled = event.project != null

        val project = event.project ?: return

        val host = project.service<PlatformLocator>().locate()

        if (host == null) {
            event.presentation.isEnabledAndVisible = false
            return
        }

        event.presentation.isEnabled = shouldBeEnabled(event)

        event.presentation.icon = host.icon
        event.presentation.text = GitLinkBundle.message("actions.${type.key}.title", host.name)
    }

    override fun getActionUpdateThread(): ActionUpdateThread {
        return ActionUpdateThread.BGT
    }
}